staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -755,3 +759,87 @@ func Test_defaultDeprecationWarning(t *testing.T) {
 		})
 	}
 }
+
+func TestBuildOpenAPIModelsForApply(t *testing.T) {
+	// This is a list of validation that we expect to work.
+	tests := []apiextensionsv1.CustomResourceValidation{
+		apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type:       "object",
+				Properties: map[string]apiextensionsv1.JSONSchemaProps{"num": {Type: "integer", Description: "v1beta1 num field"}},
+			},
+		},
+		apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type:         "",
+				XIntOrString: true,
+			},
+		},
+		apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type: "object",
+				Properties: map[string]apiextensionsv1.JSONSchemaProps{
+					"oneOf": {
+						OneOf: []apiextensionsv1.JSONSchemaProps{
+							{Type: "boolean"},
+							{Type: "string"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	staticSpec, err := getOpenAPISpecFromFile()
+	if err != nil {
+		t.Fatalf("Failed to load openapi spec: %v", err)
+	}
+
+	for _, test := range tests {
+		crd := apiextensionsv1.CustomResourceDefinition{
+			ObjectMeta: metav1.ObjectMeta{Name: "example.stable.example.com", UID: types.UID("12345")},
+			Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+				Group: "stable.example.com",
+				Names: apiextensionsv1.CustomResourceDefinitionNames{
+					Plural: "examples", Singular: "example", Kind: "Example", ShortNames: []string{"ex"}, ListKind: "ExampleList", Categories: []string{"all"},
+				},
+				Conversion:            &apiextensionsv1.CustomResourceConversion{Strategy: apiextensionsv1.NoneConverter},
+				Scope:                 apiextensionsv1.ClusterScoped,
+				PreserveUnknownFields: false,
+				Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+					{
+						Name: "v1beta1", Served: true, Storage: true,
+						Subresources: &apiextensionsv1.CustomResourceSubresources{Status: &apiextensionsv1.CustomResourceSubresourceStatus{}},
+						Schema:       &test,
+					},
+				},
+			},
+		}
+
+		if _, err := buildOpenAPIModelsForApply(staticSpec, &crd); err != nil {
+			t.Fatalf("failed to convert to apply model: %v", err)
+		}
+	}
+}
+
+func getOpenAPISpecFromFile() (*spec.Swagger, error) {
```

Comment:
I'm surprised you have to implement that, don't we already do that in a few places?

Reply:
We do, but those convert directly from the swagger file to a proto (https://github.com/kubernetes/kube-openapi/blob/master/pkg/util/proto/testing/openapi.go#L41).

We could move this to a utility file but I'm not exactly sure where we should put it.

Reply:
Yeah, ok that's fine for now, thanks! :-)